src/Features/CSharp/Portable/Snippets/CSharpForLoopSnippetProvider.cs

```diff
@@ -0,0 +1,179 @@
+
+            return forLoopSyntax;
+        }
+
+        /// <summary>
+        /// Goes through each piece of the for statement and extracts the identifiers
+        /// as well as their locations to create SnippetPlaceholder's of each.
+        /// </summary>
+        protected override ImmutableArray<SnippetPlaceholder> GetForLoopSnippetPlaceholders(SyntaxNode node, ISyntaxFacts syntaxFacts)
+        {
+            using var _ = ArrayBuilder<SnippetPlaceholder>.GetInstance(out var arrayBuilder);
+            var placeHolderBuilder = new Dictionary<string, List<int>>();
+            GetPartsOfForStatement(node, out var declaration, out var condition, out var incrementor, out var _1);
+
+            // Retrieves the placeholder present in the variable declaration as well as its location.
+            if (declaration != null)
+            {
+                var variableDeclarator = ((VariableDeclarationSyntax)declaration).Variables.FirstOrDefault();
+                if (variableDeclarator != null)
+                {
+                    var declaratorIdentifier = variableDeclarator.Identifier;
+                    placeHolderBuilder.Add(declaratorIdentifier.ToString(), new List<int>() { declaratorIdentifier.SpanStart });
+                }
+            }
+
+            // Gets the placeholders present in the left and right side of the conditional.
+            if (condition != null)
+            {
+                var conditionExpression = (BinaryExpressionSyntax)condition;
+                var left = conditionExpression.Left;
+                if (!placeHolderBuilder.TryGetValue(left.ToString(), out var incrementorList))
+                {
+                    incrementorList = new List<int>();
+                    placeHolderBuilder.Add(left.ToString(), incrementorList);
+                }
+
+                incrementorList.Add(left.SpanStart);
+
+                var right = conditionExpression.Right;
+                if (!placeHolderBuilder.TryGetValue(right.ToString(), out var lengthList))
+                {
+                    lengthList = new List<int>();
+                    placeHolderBuilder.Add(right.ToString(), lengthList);
+                }
+
+                lengthList.Add(right.SpanStart);
+            }
+
+            // Gets the placeholder present in the incrementor.
+            if (incrementor != null)
+            {
+                var operand = ((PostfixUnaryExpressionSyntax)incrementor).Operand;
+                if (!placeHolderBuilder.TryGetValue(operand.ToString(), out var incrementorList))
+                {
+                    incrementorList = new List<int>();
+                    placeHolderBuilder.Add(operand.ToString(), incrementorList);
+                }
+
+                incrementorList.Add(operand.SpanStart);
+            }
+
+            foreach (var kvp in placeHolderBuilder)
+            {
+                arrayBuilder.Add(new SnippetPlaceholder(kvp.Key, kvp.Value.ToImmutableArray()));
+            }
```

Comment:
do you need to sort this array in any special way?  or does it not matter?

Reply:
It doesn't matter because when we construct the LSP string, we create a new dictionary that has a mapping of the position it should be inserted into the string to the identifier